scheduler/scheduler.go

```diff
@@ -57,32 +57,49 @@ func (p *PortResourcePool) IsIPQualifiedForRequests(ip string, specs []PortSpec)
 // ReserveIPPort reserve an ip and port from a port pool
 func (p *PortResourcePool) ReserveIPPort(ip string, port int64) error {
 	if _, ok := p.PortBindingMap[ip]; !ok {
-		// if ip can't be found,  reserve for the default ip 0.0.0.0
-		if _, ok := p.PortBindingMap[defaultIP]; ok {
-			p.PortBindingMap[defaultIP][port] = true
-			return nil
-		}
-		// if 0.0.0.0 is not on the pool
-		if ip == defaultIP {
-			// in this case, 0.0.0.0 is not on the pool so there must be multiple ip on the host
-			// reserve all ips on the specified port
-			success := true
-			for key := range p.PortBindingMap {
-				if !p.PortBindingMap[key][port] {
-					p.PortBindingMap[key][port] = true
-					continue
-				}
-				success = false
-				break
+		// if ip can't be found and it is not 0.0.0.0,  reserve on the ghost map
+		if ip != defaultIP {
+			if _, ok := p.GhostMap[ip]; !ok {
+				p.GhostMap[ip] = map[int64]bool{}
+				logrus.Infof("Creating ghost map for IP %v", ip)
+				p.GhostMap[ip][port] = true
+				logrus.Infof("Port %v is reserved for IP %v for ghost map", port, ip)
+				return nil
 			}
-			if success {
+			if !p.GhostMap[ip][port] {
+				p.GhostMap[ip][port] = true
+				logrus.Infof("Port %v is reserved for IP %v for ghost map", port, ip)
 				return nil
 			}
+			return errors.Errorf("Port %v is already used in ip %v in ghost map", port, ip)
```

Comment:
This error message might make its way back to the user, so we cannot use the term "ghost map". Just drop the phrase "in ghost map" from the end.